src/audio/copier/copier_hifi.c

```diff
@@ -137,4 +137,374 @@ int copier_gain_set_fade_params(struct comp_dev *dev, struct dai_data *dd,
+			}
+
+			AE_SA64POS_FP(align_out, out_ptr);
+			if (rest) {
+				r_d32_l = AE_ZERO32();
+				if (rest > 1) {
+					AE_SA32X2_IC(r_d32_l, align_out, out_ptr);
+					if (rest > 2) {
+						ae_int32 tmp = AE_MOVAD32_H(r_d32_h);
+
+						AE_S32_L_XC(tmp, (ae_int32 *)out_ptr, 0);
+					}
+				} else {
+					ae_int32 tmp = AE_MOVAD32_H(r_d32_l);
+
+					AE_S32_L_XC(tmp, (ae_int32 *)out_ptr, 0);
+				}
+			}
+			samples -= nmax;
+			dst = audio_stream_wrap(&buff->stream, dst + nmax);
+		}
+		gain_params->silence_sg_count += sg_count;
+		break;
+	case TRANS_GAIN:
+		ae_f16x4 gain_env = (int16_t)(gain_params->gain_env >> I64_TO_I16_SHIFT);
+
+		gain_env = AE_ADD16S(gain_env, gain_params->init_gain);
+		while (samples) {
+			nmax = audio_stream_samples_without_wrap_s32(&buff->stream, dst);
+			out_ptr = (ae_int32x2 *)(dst);
+			in_ptr = (ae_int32x2 *)(dst);
+			AE_LA32X2POS_PC(align_in, in_ptr);
+			nmax = MIN(samples, nmax);
+			rest = nmax & 0x3;
+			for (n = 0; n < (nmax >> 2); n++) {
+				/* static gain part */
+				if (!gain_params->unity_gain) {
+					copier_load_slots_and_gain32(&in_ptr, &align_in, gain_i16,
+								     &d32_h, &d32_l);
+				} else {
+					AE_LA32X2_IC(d32_h, align_in, in_ptr);
+					AE_LA32X2_IC(d32_l, align_in, in_ptr);
+				}
+				/* quadratic fade-in part */
+				d32_h = AE_MULFP32X16X2RAS_H(d32_h, gain_env);
+				d32_h = AE_MULFP32X16X2RAS_H(d32_h, gain_env);
+				d32_l = AE_MULFP32X16X2RAS_L(d32_l, gain_env);
+				d32_l = AE_MULFP32X16X2RAS_L(d32_l, gain_env);
+				AE_SA32X2_IC(d32_h, align_out, out_ptr);
+				AE_SA32X2_IC(d32_l, align_out, out_ptr);
+
+				if (dir == ADDITION)
+					gain_env = AE_ADD16S(gain_env, gain_params->step_f16);
+				else
+					gain_env = AE_SUB16S(gain_env, gain_params->step_f16);
+			}
+
+			AE_SA64POS_FP(align_out, out_ptr);
+			if (rest) {
+				/* static gain part */
+				if (!gain_params->unity_gain) {
+					copier_load_slots_and_gain32(&in_ptr, &align_in, gain_i16,
+								     &r_d32_h, &r_d32_l);
+				} else {
+					AE_LA32X2_IC(r_d32_h, align_in, in_ptr);
+					AE_LA32X2_IC(r_d32_l, align_in, in_ptr);
+				}
+				/* quadratic fade-in part */
+				r_d32_h = AE_MULFP32X16X2RAS_H(r_d32_h, gain_env);
+				r_d32_h = AE_MULFP32X16X2RAS_H(r_d32_h, gain_env);
+				r_d32_l = AE_MULFP32X16X2RAS_L(r_d32_l, gain_env);
+				r_d32_l = AE_MULFP32X16X2RAS_L(r_d32_l, gain_env);
+
+				if (rest > 1) {
+					AE_SA32X2_IC(r_d32_l, align_out, out_ptr);
+					if (rest > 2) {
+						ae_int32 tmp = AE_MOVAD32_H(r_d32_h);
+
+						AE_S32_L_XC(tmp, (ae_int32 *)out_ptr, 0);
+					}
+				} else {
+					ae_int32 tmp = AE_MOVAD32_H(r_d32_l);
+
+					AE_S32_L_XC(tmp, (ae_int32 *)out_ptr, 0);
+				}
+			}
+			samples -= nmax;
+			dst = audio_stream_wrap(&buff->stream, dst + nmax);
+		}
+		gain_params->fade_in_sg_count += sg_count;
+
+		if (dir == ADDITION)
+			gain_params->gain_env += gain_params->step_i64 * sg_count;
+		else
+			gain_params->gain_env -= gain_params->step_i64 * sg_count;
+	}
+	return 0;
```

Comment:
the amount of code added by a gain function is staggering.

we will run out of memory very quicky at this rate.

Reply:
the dmic-based topologies are like a tree, with multiple PCM capture streams exposed to the user, each with their own gain. So adding the gain in the copier at the base of the tree does not help. There will be duplication of gain functionality due to the generation of multiple PCM paths.

So the only sensible course of action is to use a library that's shared between all gain-related modules.